File: PerformanceTests/StitchMarker/ck/regressions/ck_hs/benchmark/apply.c
/*
 * Copyright 2014 Samy Al Bahra.
 * Copyright 2014 Backtrace I/O, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyrighs
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyrighs
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <ck_hs.h>

#include <assert.h>
#include <ck_malloc.h>
#include <errno.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "../../common.h"
#include "../../../src/ck_ht_hash.h"

static ck_hs_t hs;
static char **keys;
static size_t keys_length = 0;
static size_t keys_capacity = 128;
static unsigned long global_seed;

static void *
hs_malloc(size_t r)
{

	return malloc(r);
}

static void
hs_free(void *p, size_t b, bool r)
{

	(void)b;
	(void)r;

	free(p);

	return;
}

static struct ck_malloc my_allocator = {
	.malloc = hs_malloc,
	.free = hs_free
};

static unsigned long
hs_hash(const void *object, unsigned long seed)
{
	const char *c = object;
	unsigned long h;

	h = (unsigned long)MurmurHash64A(c, strlen(c), seed);
	return h;
}

static bool
hs_compare(const void *previous, const void *compare)
{

	return strcmp(previous, compare) == 0;
}

static void
set_destroy(void)
{

	ck_hs_destroy(&hs);
	return;
}

static void
set_init(unsigned int size, unsigned int mode)
{

	if (ck_hs_init(&hs, CK_HS_MODE_OBJECT | CK_HS_MODE_SPMC | mode, hs_hash, hs_compare,
	    &my_allocator, size, global_seed) == false) {
		perror("ck_hs_init");
		exit(EXIT_FAILURE);
	}

	return;
}

static size_t
set_count(void)
{

	return ck_hs_count(&hs);
}

static bool
set_reset(void)
{

	return ck_hs_reset(&hs);
}

static void *
test_apply(void *key, void *closure)
{

	(void)key;

	return closure;
}

static void
run_test(const char *file, size_t r, unsigned int size, unsigned int mode)
{
	FILE *fp;
	char buffer[512];
	size_t i, j;
	unsigned int d = 0;
	uint64_t s, e, a, gp, agp;
	struct ck_hs_stat st;
	char **t;

	keys = malloc(sizeof(char *) * keys_capacity);
	assert(keys != NULL);

	fp = fopen(file, "r");
	assert(fp != NULL);

	while (fgets(buffer, sizeof(buffer), fp) != NULL) {
		buffer[strlen(buffer) - 1] = '\0';
		keys[keys_length++] = strdup(buffer);
		assert(keys[keys_length - 1] != NULL);

		if (keys_length == keys_capacity) {
			t = realloc(keys, sizeof(char *) * (keys_capacity *= 2));
			assert(t != NULL);
			keys = t;
		}
	}

	t = realloc(keys, sizeof(char *) * keys_length);
	assert(t != NULL);
	keys = t;

	set_init(size, mode);
	for (i = 0; i < keys_length; i++) {
		unsigned long h = CK_HS_HASH(&hs, hs_hash, keys[i]);

		if (ck_hs_get(&hs, h, keys[i]) == false) {
			if (ck_hs_put(&hs, h, keys[i]) == false)
				ck_error("ERROR: Failed get to put workload.\n");
		} else {
			d++;
		}
	}
	ck_hs_stat(&hs, &st);

	fprintf(stderr, "# %zu entries stored, %u duplicates, %u probe.\n",
	    set_count(), d, st.probe_maximum);

	a = 0;
	for (j = 0; j < r; j++) {
		if (set_reset() == false)
			ck_error("ERROR: Failed to reset hash table.\n");

		s = rdtsc();
		for (i = 0; i < keys_length; i++) {
			unsigned long h = CK_HS_HASH(&hs, hs_hash, keys[i]);

			if (ck_hs_get(&hs, h, keys[i]) == false &&
			    ck_hs_put(&hs, h, keys[i]) == false) {
				ck_error("ERROR: Failed get to put workload.\n");
			}
		}
		e = rdtsc();
		a += e - s;
	}
	gp = a / (r * keys_length);

	a = 0;
	for (j = 0; j < r; j++) {
		if (set_reset() == false)
			ck_error("ERROR: Failed to reset hash table.\n");

		s = rdtsc();
		for (i = 0; i < keys_length; i++) {
			unsigned long h = CK_HS_HASH(&hs, hs_hash, keys[i]);

			if (ck_hs_apply(&hs, h, keys[i], test_apply, (void *)keys[i]) == false)
				ck_error("ERROR: Failed in apply workload.\n");
		}
		e = rdtsc();
		a += e - s;
	}
	agp = a / (r * keys_length);

	fclose(fp);

	for (i = 0; i < keys_length; i++) {
		free(keys[i]);
	}

	printf("Get to put: %" PRIu64 " ticks\n", gp);
	printf("     Apply: %" PRIu64 " ticks\n", agp);

	free(keys);
	keys_length = 0;
	set_destroy();
	return;
}

int
main(int argc, char *argv[])
{
	unsigned int r, size;

	common_srand48((long int)time(NULL));
	if (argc < 2) {
		ck_error("Usage: ck_hs <dictionary> [<repetitions> <initial size>]\n");
	}

	r = 16;
	if (argc >= 3)
		r = atoi(argv[2]);

	size = 8;
	if (argc >= 4)
		size = atoi(argv[3]);

	global_seed = common_lrand48();
	run_test(argv[1], r, size, 0);

	printf("\n==============================================\n"
	    "Delete mode\n"
	    "==============================================\n");
	run_test(argv[1], r, size, CK_HS_MODE_DELETE);
	return 0;
}

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/apply.py
# Copyright (C) 2026 Apple Inc. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1.  Redistributions of source code must retain the above copyright
#     notice, this list of conditions and the following disclaimer.
# 2.  Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import argparse
import os
import sys
import tempfile

from .command import Command
from .diff.html_diff import HTMLDiff
from .pull_request import PullRequest

from webkitbugspy import radar, Tracker
from webkitcorepy import arguments, run, Terminal
from webkitscmpy import local, log


class Apply(Command):
    name = 'apply'
    help = 'Apply a patch attached to a bug tracker issue (for example, a proposed revert or build fix)'

    @classmethod
    def parser(cls, parser, loggers=None):
        parser.add_argument(
            'issue', type=str,
            help='Issue the patch is attached to (an issue URL, rdar://<id>, or a bare radar id)',
        )
        parser.add_argument(
            'name', type=str, default=None, nargs='?',
            help='Name of the attached patch to apply; if omitted, choose from the issue\'s patches',
        )
        parser.add_argument(
            '--no-commit', default=True, action='store_false', dest='commit',
            help='Apply the patch to the working tree only, instead of committing it with `git am`',
        )

    @classmethod
    def resolve_issue(cls, string):
        '''Resolve a tracker link, URL, or bare radar id to a webkitbugspy issue.'''
        issue = Tracker.from_string(string)
        if issue:
            return issue
        # from_string rejects a bare number; a bare id is only meaningful for radar, so resolve one
        # against the radar tracker -- digits-only, so a malformed link is never coerced into an id.
        stripped = string.strip()
        if stripped.isdigit():
            for tracker in Tracker._trackers:
                if isinstance(tracker, radar.Tracker):
                    return tracker.issue(int(stripped))
        return None

    @classmethod
    def review(cls, patch, repository, alternatives=False):
        '''Open the patch as a blocking HTML diff and return whether the user accepted it. Closing the
        diff or pressing Enter accepts it; answering 'n' rejects it.'''
        content = patch.contents()
        if content is None:
            sys.stderr.write("Could not download '{}'\n".format(patch.name))
            return False
        text = content if isinstance(content, str) else content.decode('utf-8', errors='replace')
        dismiss = 'pick a different patch' if alternatives else 'cancel'
        sys.stdout.write("Reviewing '{}' -- close the diff or press Enter to apply it, or enter 'n' to {}.\n".format(patch.name, dismiss))
        with HTMLDiff(block=True, repository=repository) as diff_viewer:
            diff_viewer.add_lines(text.splitlines())
        return diff_viewer.accepted

    @classmethod
    def select_patch(cls, patches, name=None, repository=None, interactive=False):
        '''Return the patch to apply, or None. With `name`, the exact match. Otherwise non-interactive
        returns the first patch; interactive shows a menu in the order the tracker lists the patches
        (skipped when there is only one), defaulting to the first, and reviews the chosen patch's diff
        -- closing or accepting the diff selects it, while answering 'n' returns to the menu.'''
        by_name = {patch.name: patch for patch in patches}
        if name:
            patch = by_name.get(name)
            if patch is None or not interactive:
                return patch
            return patch if cls.review(patch, repository) else None
        options = [patch.name for patch in patches]
        if not interactive:
            return patches[0]
        while True:
            choice = options[0] if len(options) == 1 else Terminal.choose(
                'Which patch would you like to apply?',
                options=options, default=options[0], numbered=True,
            )
            if cls.review(by_name[choice], repository, alternatives=len(options) > 1):
                return by_name[choice]
            if len(options) == 1:
                return None

    @classmethod
    def edit_commit_message(cls, repository):
        '''Amend the just-applied commit without a message flag, so git opens its message in the editor.
        No --date, so `git am`'s preserved author date survives.'''
        if run([repository.executable(), 'commit', '--amend'], cwd=repository.root_path).returncode:
            sys.stderr.write("Applied the patch, but couldn't open the commit message for editing; it keeps the patch's message.\n")

    @classmethod
    def offer_pull_request(cls, repository, **kwargs):
        '''Offer to open a pull request for the just-applied commit, running pull-request in-process
        with its default arguments. LoggingGroup supplies the shared verbosity args its main reads,
        mirroring how the top-level parser builds each sub-command.'''
        if Terminal.choose('Would you like to create a pull request?', default='Yes') != 'Yes':
            return
        parser = argparse.ArgumentParser()
        arguments.LoggingGroup(parser)
        PullRequest.parser(parser)
        PullRequest.main(parser.parse_args([]), repository, **kwargs)

    @classmethod
    def main(cls, args, repository, **kwargs):
        if not isinstance(repository, local.Git):
            sys.stderr.write("Can only '{}' on a native Git repository\n".format(cls.name))
            return 1
        if repository.modified():
            sys.stderr.write('Please commit or stash your changes before applying a patch\n')
            return 1

        issue = cls.resolve_issue(args.issue)
        if not issue:
            sys.stderr.write("Could not resolve '{}' to an issue\n".format(args.issue))
            return 1

        patches = issue.patches
        if patches is None:
            sys.stderr.write('{} does not support patch attachments\n'.format(issue.tracker.NAME))
            return 1
        if not patches:
            sys.stderr.write('No patches are attached to {}\n'.format(issue.link))
            return 1

        interactive = Terminal.isatty(sys.stdout) and Terminal.isatty(sys.stdin)
        patch = cls.select_patch(patches, name=args.name, repository=repository, interactive=interactive)
        if not patch:
            if args.name and not any(candidate.name == args.name for candidate in patches):
                sys.stderr.write("No '{}' patch is attached to {}\n".format(args.name, issue.link))
            else:
                sys.stderr.write('No patch applied\n')
            return 1

        # The patch is issue-attachment content -- whoever can attach to the issue controls these
        # bytes, applied via git am/apply: the same trust as applying any colleague's patch.
        content = patch.contents()
        if content is None:
            sys.stderr.write("Could not download '{}' from {}\n".format(patch.name, issue.link))
            return 1
        if not isinstance(content, bytes):
            content = content.encode('utf-8')
        if not content.strip():
            sys.stderr.write("'{}' on {} is empty\n".format(patch.name, issue.link))
            return 1

        handle, patch_path = tempfile.mkstemp(suffix='.patch')
        try:
            with os.fdopen(handle, 'wb') as patch_file:
                patch_file.write(content)

            if args.commit:
                result = run([repository.executable(), 'am', patch_path], cwd=repository.root_path)
            else:
                result = run([repository.executable(), 'apply', '--3way', patch_path], cwd=repository.root_path)

            if result.returncode:
                sys.stderr.write("Failed to apply '{}' from {}\n".format(patch.name, issue.link))
                if args.commit:
                    run([repository.executable(), 'am', '--abort'], cwd=repository.root_path, capture_output=True)
                    sys.stderr.write('Re-run with --no-commit to apply it to the working tree and resolve conflicts by hand.\n')
                else:
                    sys.stderr.write('Conflicting hunks were left as conflict markers / .rej files; `git checkout -- .` to discard.\n')
                return 1
        finally:
            os.remove(patch_path)

        if args.commit and interactive:
            cls.edit_commit_message(repository)
            cls.offer_pull_request(repository, **kwargs)

        log.info("Applied '{}' from {}".format(patch.name, issue.link))
        return 0
